pkg/operator/ceph/cluster/mon/health.go

```diff
@@ -366,3 +314,133 @@ func removeMonitorFromQuorum(context *clusterd.Context, clusterName, name string
+}
+
+func (c *Cluster) findInvalidMonitorPlacement(desiredMonCount int) (*NodeUsage, error) {
+	nodeZones, err := c.getNodeMonUsage()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get node monitor usage. %+v", err)
+	}
+
+	// compute two helpful global flags:
+	//  - does an empty zone exist
+	//  - does an empty node exist
+	emptyZone := false
+	emptyNode := false
+	for zi := range nodeZones {
+		monFoundInZone := false
+		for ni := range nodeZones[zi] {
+			nodeUsage := &nodeZones[zi][ni]
+			if nodeUsage.MonCount > 0 {
+				monFoundInZone = true
+			} else if nodeUsage.MonValid {
+				// only consider valid nodes since this flag determines if a mon
+				// may be scheduled on to a new node.
+				emptyNode = true
+			}
+		}
+		if !monFoundInZone {
+			emptyZone = true
+		}
+	}
+
+	logger.Infof("rebalance: desired mon count: %d, empty zone found: %t, empty node found: %t",
+		desiredMonCount, emptyZone, emptyNode)
+
+	for zi := range nodeZones {
+		zoneMonCount := 0
+		var nodeChoice *NodeUsage = nil
+
+		// for each node consider two cases: too many monitors are running on a
+		// node, or monitors are running on invalid nodes.
+		for ni := range nodeZones[zi] {
+			nodeUsage := &nodeZones[zi][ni]
+			zoneMonCount += nodeUsage.MonCount
+
+			// if this node has too many monitors, and an underused node exists,
+			// then consider moving the monitor. note that this check is
+			// independent of the setting `AllowMultiplePerNode` since we also
+			// want to avoid in general keeping multiple monitors on one node.
+			if nodeUsage.MonCount > 1 && emptyNode {
+				logger.Infof("rebalance: chose overloaded node %s with %d mons",
+					nodeUsage.Node.Name, nodeUsage.MonCount)
+				nodeChoice = nodeUsage
+				break
+			}
+
+			// check for mons on invalid nodes. but reschedule pod only when it
+			// is invalid for reasons other than schedulability which should not
+			// affect pods that are already scheduled.
+			if nodeUsage.MonCount > 0 && !nodeUsage.MonValid {
+				placement := cephv1.GetMonPlacement(c.spec.Placement)
+				placement.Tolerations = append(placement.Tolerations,
+					v1.Toleration{
+						Key:      scheduler.TaintNodeUnschedulable,
+						Effect:   "",
+						Operator: v1.TolerationOpExists,
+					})
+				valid, err := k8sutil.ValidNodeNoSched(*nodeUsage.Node, placement)
+				if err != nil {
+					logger.Warningf("rebalance: failed to validate node %s %+v",
+						nodeUsage.Node.Name, err)
+				} else if !valid {
+					logger.Infof("rebalance: chose invalid node %s",
+						nodeUsage.Node.Name)
+					nodeChoice = nodeUsage
+					break
+				}
+			}
+		}
+
+		// if we didn't already select a node with a monitor for rescheduling,
+		// then consider that a zone may be overloaded and can be rebalanced.
+		// this case occurs if the zone has more than 1 monitor, and some other
+		// zone exists without any monitors.
+		if nodeChoice == nil && zoneMonCount > 1 && emptyZone {
+			for ni := range nodeZones[zi] {
```

Comment:
This for loop is not captured in unit test coverage.

Reply:
nice catch. does the the normal `make check` give me coverage? I hadn't seen this but maybe didn't know I could look for it.

Reply:
I found it by turning coverage highlighting on in VSCode. I'm really not sure what it uses under the covers.

Reply:
oh, thanks. yeh i found this `go test -coverprofile=coverage.out; go tool cover -func=coverage.out`  opens a browser with the coverage results.

Reply:
fixed